### PR TITLE
Add proxy client certs to master config

### PR DESCRIPTION
**Pull request by abutcher**
Add proxy client certs to the master config. The upgrade playbook upgrades the base package first so that we can run `create-master-certs` while generating the rest of the configs. 

~~<3.1 installs will be running `create-master-certs` every run since they'll be missing `master.proxy-client.crt` - it's safe to do with `--overwrite=false` but we could combine version based lists of certs when we have the version fact to get around it...~~

For #685.

@sdodson @detiber @wshearn @brenton @liggitt

### CODE REVIEW

**Comment by detiber** [comment]
@abutcher glad you mentioned that about <3.1 installs

we can avoid that using a template...  will make the comment inline.

File: playbooks/common/openshift-master/config.yml [hunk 1]
```diff
@@ -122,6 +122,7 @@
       openshift_master_certs_no_etcd:
       - admin.crt
       - master.kubelet-client.crt
+      - master.proxy-client.crt
```
Comment (detiber) [inline playbooks/common/openshift-master/config.yml]: this template should do the trick, of course I'm making up a variable containing the OpenShift version (not sure if one already exists):
- "{{ 'master.proxy.client.crt' if openshift_version >= 3.1 else omit }}"

The omit keyword should cause that loop iteration to be a noop when the version is less than 3.1. of course, using the version comparison filter is probably the right way to compare as well.

Comment (abutcher) [inline playbooks/common/openshift-master/config.yml]: @detiber Awesome - throwing this together

File: roles/openshift_master/templates/master.yaml.v1.j2 [hunk 2]
```diff
@@ -74,6 +74,9 @@ kubernetesMasterConfig:
   masterCount: 1
   masterIP: ""
   podEvictionTimeout: ""
+  proxyClientInfo:
```
Comment (liggitt) [inline roles/openshift_master/templates/master.yaml.v1.j2]: does this need to be conditionalized for version as well? it'll get ignored pre-3.1, but just for hygiene...

Comment (sdodson) [inline roles/openshift_master/templates/master.yaml.v1.j2]: Thus far we've not conditionalized bits in the template if we knew they'd be ignored under older versions. I can see how that'd be confusing to users but we mostly assume that if they're running the latest ansible code they're probably going to be running the latest product code.

**Comment by abutcher** [comment]
@detiber Check that out. I also changed `master_certs_missing` from a list to a bool which we should probably do with the other `*cert_missing` facts?

File: roles/openshift_master/templates/master.yaml.v1.j2 (hunk 2 above)
Comment (abutcher) [inline roles/openshift_master/templates/master.yaml.v1.j2]: We also need something to update version fact right after we install packages so that we can use it on first run.

Comment (detiber) [inline roles/openshift_master/templates/master.yaml.v1.j2]: I agree with @sdodson, I'd like to avoid adding logic to the templates for items that are ignored. Longer term the templates will be going away in favor of a better method to manage configuration.

Comment (detiber) [inline roles/openshift_master/templates/master.yaml.v1.j2]: @abutcher The easiest way will be to conditionally run openshift_facts to set the version if the state of the installation task is successful.

**Comment by detiber** [comment]
@abutcher :+1: on making the other cert missing vars into bools

File: playbooks/common/openshift-master/config.yml [hunk 3]
```diff
@@ -122,6 +122,7 @@
       openshift_master_certs_no_etcd:
       - admin.crt
       - master.kubelet-client.crt
+      - "{{ 'master.proxy-client.crt' if (openshift.common.version != '' and (openshift.common.version | version_compare('1.0.6', '>') and openshift.common.version | version_compare('3.0','<') ) or openshift.common.version | version_compare('3.0.2','>')) else omit }}"
```
Comment (detiber) [inline playbooks/common/openshift-master/config.yml]: This conditional is a bit hard to follow...  might want to use a combination of deployment_type and version instead of testing > 1.0.6 and < 3.0

either way, it is definitely going to be hairy...  might want to use set_fact to set a variable to use to conditionally omit the master.proxy-client.crt.

Comment (abutcher) [inline playbooks/common/openshift-master/config.yml]: Yeah it's ugly let me see what I can do

**Comment by abutcher** [comment]
@detiber Little bit cleaner...

File: playbooks/common/openshift-master/config.yml [hunk 4]
```diff
@@ -138,9 +141,9 @@
     with_items: openshift_master_certs
     register: g_master_cert_stat_result
   - set_fact:
-      master_certs_missing: "{{ g_master_cert_stat_result.results
+      master_certs_missing: "{{ False in (g_master_cert_stat_result.results
                                 | map(attribute='stat.exists')
-                                | list | intersect([false])}}"
+                                | list ) }}"
```
Comment (detiber) [inline playbooks/common/openshift-master/config.yml]: Nice! much cleaner.

File: playbooks/common/openshift-master/config.yml [hunk 5]
```diff
@@ -119,9 +119,11 @@
   hosts: oo_masters_to_config
   tasks:
   - set_fact:
+      include_proxy_client_cert: "{{ (openshift.common.version | version_compare('1.0.6', '>')) if openshift.common.deployment_type == 'origin' else (openshift.common.version | version_compare('3.0.2', '>')) }}"
       openshift_master_certs_no_etcd:
       - admin.crt
       - master.kubelet-client.crt
+      - "{{ 'master.proxy-client.crt' if include_proxy_client_cert else omit }}"
```
Comment (detiber) [inline playbooks/common/openshift-master/config.yml]: :heart:

File: roles/openshift_master_certificates/tasks/main.yml [hunk 6]
```diff
@@ -20,6 +20,8 @@
     - admin.kubeconfig
     - master.kubelet-client.crt
     - master.kubelet-client.key
+    - "{{ 'master.proxy-client.crt' if openshift.master.include_proxy_client_cert else omit }}"
+    - "{{ 'master.proxy-client.key' if openshift.master.include_proxy_client_cert else omit }}"
```
Comment (detiber) [inline roles/openshift_master_certificates/tasks/main.yml]: Have you tested this? I wouldn't expect this to be available unless you set it through openshift_facts and not just setting it with set_fact in the master role.

Comment (abutcher) [inline roles/openshift_master_certificates/tasks/main.yml]: Very right - it succeeds but only because `masters_needing_certs` is empty on my single node setup.

Comment (abutcher) [inline roles/openshift_master_certificates/tasks/main.yml]: @detiber Are you more partial to moving things like this into facts or setting them where needed?

Comment (detiber) [inline roles/openshift_master_certificates/tasks/main.yml]: @abutcher anything that is referenced across roles should be persisted through openshift_facts.

**Comment by abutcher** [comment]
@detiber Bumped. Only not obvious thing in the latest commit is that the version is defaulted to `None` so that we can use it in comparisons, however, this means that the fact will always be `True` on initial installation. This is okay for the certificates because they will all be missing at first but if this fact saw wider use it might make someone pull their hair out.

```
>>> from ansible.runner.filter_plugins.core import version_compare
>>> version_compare(None, '1.0.6', '>')
True
>>>
```

**Comment by detiber** [comment]
@abutcher we'll probably want to test an installation done with the current master branch using 3.0.2, and re-running using this branch to make sure we handle that use case properly. Otherwise, the approach sounds good to me.

**Comment by abutcher** [comment]
@detiber Tested master 3.0.2 install and a re-run with this branch. The facts are already being regathered after the first package install so this fact should always be correct.

**Comment by detiber** [comment]
@abutcher outstanding! ship it!

**Comment by brenton** [comment]
Should this be merged now?

**Comment by abutcher** [comment]
I moved the BZ to QE - we can wait for their verification.

**Comment by abutcher** [comment]
@brenton This one has been verified by QE.
